File: interview/questions/advanceArray.java
package interview.questions;

import java.util.HashMap;

/* Data Structure that performs set(), get(), setAll() in O(1)
https://stackoverflow.com/questions/72591504/data-structure-that-performs-set-get-setall-in-o1
Implement a Data Structure that capable of getting an array of numbers and have init([array]), setAll(value), getIndex(index), and setIndex(index,value) functions usage all with O(1) complexity
setAll(value)->setAll(5)->[5,5,5,5,5]
getIndex(index)->value->getIndex(2)=>3
setIndex(index,value)->setIndex(0,12)->[12,...] 
*/

public class advanceArray {
    boolean setAllStatus = false;
    HashMap<Integer, Integer> hasMap = new HashMap<>();
    int setAllValue = 0;
    int count = 0;
    public void set(int key, int value) {
        hasMap.put(key, value);
    }
    public int get(int key) {
        if (setAllStatus) {
            if (hasMap.get(key) != null) {
                if (count == hasMap.size()) {
                    return setAllValue;
                } else {
                    // do something
                }
            } else {
                throw new NullPointerException();
            }
        } else {
            if (hasMap.get(key) == null) {
                throw new NullPointerException();
            } else {
                return hasMap.get(key);
            }
        }
        return key;
    }
    public void setAll(int value) {
        setAllStatus = true;
        setAllValue = value;
        count = hasMap.size();
    }
    public static void main(String[] args) {
        advanceArray m = new advanceArray();
        m.set(1, 4);
        m.set(4, 5);
        System.out.println(m.get(4)); // 5
        m.setAll(6);
        System.out.println(m.get(4)); // 6
        m.set(8, 7);
        System.out.println(m.get(8)); // 7
    }
}

// public class SingleValueMap<K, V> {
//     private Map<K, V> map = new HashMap<>();
//     private V commonValue;
    
//     public void setAll(V newValue) {
//         this.commonValue = newValue;
//     }
    
//     public void add(K key) {
//         map.put(key, commonValue);
//     }
    
//     public void add(K key, V newValue) {
//         setAll(newValue);
//         map.put(key, commonValue);
//     }
    
//     public V get(K key) {
//         if (!map.containsKey(key)) throw new NoSuchElementException();
        
//         return commonValue;
//     }
// }

// public class SingleValueSet<K, V> {
//     private Set<K> set = new HashSet<>();
//     private V commonValue;
    
//     public void setAll(V newValue) {
//         this.commonValue = newValue;
//     }
    
//     public void add(K key) {
//         set.add(key);
//     }
    
//     public void add(K key, V newValue) {
//         setAll(newValue);
//         set.add(key);
//     }
    
//     public V get(K key) {
//         if (!set.contains(key)) throw new NoSuchElementException();
        
//         return commonValue;
//     }
// }

File: interview/questions/advanceArray.js
let arr = [1, 2, 3, 4, 5];
/* Data Structure that performs set(), get(), setAll() in O(1)
Implement a Data Structure that capable of getting an array of numbers and have init([array]), setAll(value), getIndex(index), and setIndex(index,value) functions usage all with O(1) complexity
setAll(value)->setAll(5)->[5,5,5,5,5]
getIndex(index)->value->getIndex(2)=>3
setIndex(index,value)->setIndex(0,12)->[12,...] 
*/
class advanceDS {
  constructor() {
    this.data = {}; // Object for storing key-value pairs
    this.allValues = []; // Array to store all values
    this.allValueIndex = {}; // Object to map value indices
  }

  set(key, value) {
    if (!this.data.hasOwnProperty(key)) {
      this.allValues.push(value); // Add value to the array
      this.allValueIndex[key] = this.allValues.length - 1; // Store value index
    } else {
      const index = this.allValueIndex[key]; // Get index of existing value
      this.allValues[index] = value; // Update value in the array
    }
    this.data[key] = value; // Update key-value pair in the object
  }

  get(key) {
    if (this.data.hasOwnProperty(key)) {
      return this.data[key]; // Retrieve value from the object
    }
    return undefined; // Return undefined if key doesn't exist
  }

  setAll(value) {
    this.allValues = new Array(Object.keys(this.data).length).fill(value); // Update all values in the array
    Object.keys(this.data).forEach((key, index) => {
      this.data[key] = value; // Update all key-value pairs in the object
      this.allValueIndex[key] = index; // Update value indices
    });
  }
}

// Instantiate Constant Advance Data Structure
const ds = new advanceDS();

// Test set() and get()
ds.set("key1", "value1");
console.log(ds.get("key1")); // Output: value1

// Test set() to update existing key
ds.set("key1", "updatedValue1");
console.log(ds.get("key1")); // Output: updatedValue1

// Test setAll() and get() after setAll()
ds.set("key2", "value2");
ds.setAll("newValue");
console.log(ds.get("key1")); // Output: newValue
console.log(ds.get("key2")); // Output: newValue

// Adding key-value pairs
ds.set("a", 10);
ds.set("b", 20);
ds.set("c", 30);

// Retrieving values
console.log(ds.get("a")); // Output: 10
console.log(ds.get("b")); // Output: 20
console.log(ds.get("c")); // Output: 30

// Setting new values
ds.set("a", 100);
ds.set("d", 40);

// Retrieving updated values
console.log(ds.get("a")); // Output: 100
console.log(ds.get("d")); // Output: 40

// Setting all values to 50
ds.setAll(50);

// Retrieving all updated values
console.log(ds.get("a")); // Output: 50
console.log(ds.get("b")); // Output: 50
console.log(ds.get("c")); // Output: 50
console.log(ds.get("d")); // Output: 50

// Retrieving updated values
ds.set("a", 100);
ds.set("d", 40);

console.log("ds", ds);
